flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/strategy/PipelinedRegionSchedulingStrategy.java

```diff
@@ -226,44 +228,71 @@ private void maybeScheduleRegions(final Set<SchedulingPipelinedRegion> regions)
                 SchedulingStrategyUtils.sortPipelinedRegionsInTopologicalOrder(
                         schedulingTopology, regions);
 
-        final Map<ConsumedPartitionGroup, Boolean> consumableStatusCache = new HashMap<>();
-        final Set<SchedulingPipelinedRegion> downstreamSchedulableRegions = new HashSet<>();
-        for (SchedulingPipelinedRegion region : regionsSorted) {
-            if (maybeScheduleRegion(region, consumableStatusCache)) {
-                downstreamSchedulableRegions.addAll(
-                        consumedPartitionGroupsOfRegion.getOrDefault(region, Collections.emptySet())
-                                .stream()
-                                .flatMap(
-                                        consumedPartitionGroups ->
-                                                partitionGroupConsumerRegions
-                                                        .getOrDefault(
-                                                                consumedPartitionGroups,
-                                                                Collections.emptySet())
-                                                        .stream())
-                                .collect(Collectors.toSet()));
-            }
+        List<SchedulingPipelinedRegion> regionsToSchedule = getAllSchedulableRegions(regionsSorted);
+
+        if (!regionsToSchedule.isEmpty()) {
+            // schedule regions
+            regionsToSchedule.forEach(this::scheduleRegion);
         }
+    }
 
-        if (!downstreamSchedulableRegions.isEmpty()) {
-            maybeScheduleRegions(downstreamSchedulableRegions);
+    private List<SchedulingPipelinedRegion> getAllSchedulableRegions(
+            List<SchedulingPipelinedRegion> regionsSorted) {
+        final List<SchedulingPipelinedRegion> regionsToSchedule = new LinkedList<>();
+        final Queue<SchedulingPipelinedRegion> regionQueue = new LinkedList<>(regionsSorted);
+        while (!regionQueue.isEmpty()) {
+            final Map<ConsumedPartitionGroup, Boolean> consumableStatusCache = new HashMap<>();
+            final Set<ConsumedPartitionGroup> consumedPartitionGroupCache = new HashSet<>();
+            final Set<SchedulingPipelinedRegion> addedToQueue = new HashSet<>();
+            final int currentSize = regionQueue.size();
+            for (int num = 0; num < currentSize; num++) {
```

Comment:
IIUC, the reason we read the queue in batches is that we need to reset the above caches between batches. We should explicitly comment this. Moreover, it might be more straightforward to have a method returns the next batch of regions, rather than using a queue.

Reply:
good idea, changed to this mode.